### PR TITLE
fix #25: Support ARM64 and other platforms

### LINKED ISSUE #25: Support linux arm64/aarch64 architecture
The releases for arm64/aarch64 are already provided on the Protobuf releases page https://github.com/protocolbuffers/protobuf/releases and thus can easily be supported

### PR DESCRIPTION
**Pull request by adamchalmers**
I've added unit tests, so hopefully this works :) If you approve this PR I would appreciate cutting a new release afterwards, as there's a number of projects which would like to run this on ARM.

### CODE REVIEW

**Comment by adamchalmers** [comment]
Hmm, don't merge this, it doesn't properly detect these platforms yet.

**Comment by adamchalmers** [comment]
Actually, my test wasn't actually running on those weird platforms, it was using cross-compilation. I think setup-protoc should probably let users override `os.arch()` and specify their own arch/plat in an optional Github Action input if they choose to. But I'll open a separate PR for that.

**Comment by adamchalmers** [comment]
@MatteoPologruto @per1234 Hi! Could I please get a review on this?

File: src/installer.ts [hunk 1]
```diff
@@ -114,7 +114,48 @@ async function downloadRelease(version: string): Promise<string> {
   return await tc.cacheDir(extPath, "protoc", version);
 }
 
-function getFileName(version: string): string {
+/**
+ *
+ * @param osArch - A string identifying the operating system platform for which the Node.js binary was compiled.
+ * See https://nodejs.org/api/os.html#osplatform for possible values.
```
Comment (per1234) [inline src/installer.ts:120]:
```suggestion
 * @param osArch - A string identifying operating system CPU architecture for which the Node.js binary was compiled.
 * See https://nodejs.org/api/os.html#osarch for possible values.
```

The description and reference URL were for the operating system platform/`os.platform()`, whereas this parameter is the architecture/`os.arch()`.

File: src/installer.ts [hunk 2]
```diff
@@ -114,7 +114,48 @@ async function downloadRelease(version: string): Promise<string> {
...
+}
+
+/**
+ * Returns the filename of the protobuf compiler.
+ *
+ * @param version - The version to download
+ * @param osPlat - The operating system platform for which the Node.js binary was compiled.
+ * See https://nodejs.org/api/os.html#osplatform for more.
+ * @param osArch - The operating system CPU architecture for which the Node.js binary was compiled.
+ * See https://nodejs.org/api/os.html#osplatform for more.
```
Comment (per1234) [inline src/installer.ts:150]:
```suggestion
 * @param osArch - The operating system CPU architecture for which the Node.js binary was compiled.
 * See https://nodejs.org/api/os.html#osarch for more.
```


The description and reference URL were for the operating system platform/`os.platform()`, whereas this parameter is the architecture/`os.arch()`.

**Review by per1234 — CHANGES_REQUESTED** [review]

File: __tests__/main.test.ts [hunk 3]
```diff
@@ -14,6 +14,24 @@ process.env["RUNNER_TEMP"] = tempDir;
...
+describe("filename tests", () => {
+  const tests = [
+    ["protoc-3.20.2-linux-x86_32.zip", "linux", ""],
+    ["protoc-3.20.2-linux-x86_64.zip", "linux", "x64"],
+    ["protoc-3.20.2-linux-aarch_64.zip", "linux", "arm64"],
+    ["protoc-3.20.2-linux-ppcle_64.zip", "linux", "ppc64"],
+    ["protoc-3.20.2-linux-s390_64.zip", "linux", "s390x"],
+    ["protoc-3.20.2-osx-aarch_64.zip", "darwin", "arm64"],
+    ["protoc-3.20.2-osx-x86_64.zip", "darwin", "x64"]
+  ];
```
Comment (alessio-perugini) [inline __tests__/main.test.ts:26]:
```suggestion
    ["protoc-3.20.2-osx-x86_64.zip", "darwin", "x64"],
    ["protoc-3.20.2-win64.zip", "win32", "x64"],
    ["protoc-3.20.2-win32.zip", "win32", "x32"],
  ];
```
Do you mind adding this 2 test cases for windows?

**Review by alessio-perugini — CHANGES_REQUESTED** [review]
Thank you for your contribution! Do you mind addressing the change requested? We'd really like to merge this :nerd_face:

**Comment by adamchalmers** [comment]
Of course, I applied all suggestions.

**Comment by alessio-perugini** [comment]
> I can't get the `prettier` check to pass, despite running `npm format`. I assume I need some particular version of prettier installed? Or some particular npm/node version?

Might be related to this:
``` 
    ["protoc-3.20.2-osx-x86_64.zip", "darwin", "x64"][
      ("protoc-3.20.2-osx-x86_64.zip", "darwin", "x64")
    ],
 ```
should be:
````
    ["protoc-3.20.2-osx-x86_64.zip", "darwin", "x64"],
    ["protoc-3.20.2-osx-x86_64.zip", "darwin", "x64"],
    ["protoc-3.20.2-win64.zip", "win32", "x64"],
  ````

**Comment by adamchalmers** [comment]
OK, formatter fixed.

**Review by alessio-perugini — APPROVED** [review]
LGTM! :rocket:

**Comment by adamchalmers** [comment]
Nice! I can't merge until @per1234 changes their review status, or an admin merges this for me.

**Comment by alessio-perugini** [comment]
@adamchalmers we forgot to run `npm run build`. Can you run it and add the changes in a new commit? :pray:

**Comment by adamchalmers** [comment]
Done, I got one error from `npm run build`

```
$ npm run build

> setup-protoc-action@0.0.0 build
> tsc

node_modules/keyv/src/index.d.ts:60:32 - error TS2304: Cannot find name 'AsyncGenerator'.

60  iterator(namespace?: string): AsyncGenerator<any, void, any>;
                                  ~~~~~~~~~~~~~~


Found 1 error.
```

**Comment by alessio-perugini** [comment]
> Done, I got one error from `npm run build`
> 
> ```
> $ npm run build
> 
> > setup-protoc-action@0.0.0 build
> > tsc
> 
> node_modules/keyv/src/index.d.ts:60:32 - error TS2304: Cannot find name 'AsyncGenerator'.
> 
> 60  iterator(namespace?: string): AsyncGenerator<any, void, any>;
>                                   ~~~~~~~~~~~~~~
> 
> 
> Found 1 error.
> ```

It was something related to typescript version. If you rebase this on master and run again `npm run build` we should fix the issue. We're almost there :muscle:

**Comment by adamchalmers** [comment]
Awesome, thanks for shepherding this through, Alessio.

**Pull request merged**